### DISCUSSION TITLE
Naming a simplified version of `conditional`

### DISCUSSION THREAD

**Discussion by cjquines** [remeda/remeda#824] [Announcements]
## Proposed function

See #777 for the PR, which proposes a simplified version of `conditional` with at most two cases. The current API, with `functionName` being the function, looks like:

```ts
functionName(data, predicate, onTrue);
functionName(data, predicate, { onTrue, onFalse });

functionName(42, isNumber, (value) => value + 1); // => 43
functionName("", isNumber, (value) => value + 1); // => ""

functionName(
  42,
  (x) => x > 3,
  { onTrue: add(1), onFalse: multiply(2) }
); // => 43

// Replacement for Lodash / Ramda `defaultTo`:
function defaultTo<T>(data?: T | null, defaultValue: T): T {
  return functionName(data, isNullish, constant(defaultValue));
}
```

## Options

We're bikeshedding the name.

* `branch`, as in "branch on this predicate being true". This doesn't have any precedent in other utility libraries I've seen. This could be a positive, in that it brings attention to the fact that it's different.

```ts
pipe(
  data,
  branch(isNullish, constant(42)),
  branch(
    (x) => x > 3,
    { onTrue: add(1), onFalse: multiply(2) }
  ),
); // => 43
```

* `ifElse`. This is the name of the corresponding functions in [Ramda](https://ramdajs.com/docs/#ifElse) and [ts-belt](https://mobily.github.io/ts-belt/api/function#ifelse), at least in the case with both the `onTrue` and `onFalse` functions. Note that this could read weird in the case with only the `onTrue` though.

```ts
pipe(
  data,
  ifElse(isNullish, constant(42)),
  ifElse(
    (x) => x > 3,
    { onTrue: add(1), onFalse: multiply(2) }
  ),
); // => 43
```

* `when`. This is the name of the corresponding functions in [Ramda](https://ramdajs.com/docs/#when) and [ts-belt](https://mobily.github.io/ts-belt/api/function#when), at least in the case with only the `onTrue` function. Note that this could read weird in the case with both `onTrue` and `onFalse` though.

```ts
pipe(
  data,
  when(isNullish, constant(42)),
  when(
    (x) => x > 3,
    { onTrue: add(1), onFalse: multiply(2) }
  ),
); // => 43
```

## Why one function?

It might seem weird to use the same function for the case with only `onTrue`, and the case with both `onTrue` and `onFalse`. The main reason is simplicity: this lets us unite the implementations (and typing) for both, as they're similar. I could be convinced to split this into two functions if there's a compelling argument.

**Comment by eranhirsch** [comment]
A little more context and "behind the scenes" on how this function came to be: I originally called this `branch` because it allows you to split a pipe into 2 pipes (using the `piped` utility), with the branch call being the point where the pipe "branches out": e.g.

```ts
const result = pipe(
  map(...),
  filter(...),
  groupBy(...),
  // etc...
  branch(<condition>, {
    onTrue: piped(
      map(...),
      filter(...),
      groupBy(...),
      // etc...
    ),
    onFalse: piped(
      map(...),
      filter(...),
      groupBy(...),
      // etc...
    ),
  }),
  // From this point, the branches are merged back
  map(...),
  filter(...),
  groupBy(...),
  // etc...
);  
```

Under the hood, this function is simply a wrapper around the ternary operator (`?:`) with the correct typing to allow proper narrowing and exclusion so that the types make sense when the condition is a **type predicate**.
